draft: jsonrpc blogpost (#984)
deltachat/deltachat-pages: _posts/2024-11-10-why-jsonrpc-bindings-exist.md

This really speed up desktop and made it more responsive.

### Easier to use than cffi over process boundaries

JSON-RPC requires no linking and is transport independent.
At the moment 3 transport implementations exist (electron ipc, stdio, websocket) and it is easy to create new ones.

You could even use the new [webxdc realtime api](https://webxdc.org/docs/spec/joinRealtimeChannel.html) to connect to a remote Delta Chat instance on another computer, similar to the idea of implementing some remote desktop webxdc app.
The webxdc realtime api is also an interesting topic in of its own, read the its [announcement blog-post](./2024-11-17-webxdc-realtime) to learn more.

> ⚠️ If you use websocket, be sure to use the encrypted `wss://` variant, because plain websockets `ws://` are as unencrypted as `http`

## What's Next?

Currently we only have autogeneration for the typescript bindings,
it would be great to also have autogenerated bindings for atleast Swift, Java and Python.
Also the documentation should be improved.
The code documentation part can easily be done by copying the great documentation from the cffi and adapting it to fit to the jsonrpc api

> (this is a good-first-issue, if your dear reader want to help, improving documentation easy to get into and is always welcome).

Reviewer: This was a bit difficult to read, maybe suggestion works? Otherwise maybe re-word?
```suggestion
(this is a good-first-issue, if you, dear reader, want to help; improving documentation is easy to get into and is always welcome).
```

```c
#define DC_CONNECTIVITY_NOT_CONNECTED        1000
#define DC_CONNECTIVITY_CONNECTING           2000
#define DC_CONNECTIVITY_WORKING              3000
#define DC_CONNECTIVITY_CONNECTED            4000
int dc_get_connectivity(dc_context_t* context);

typedef struct _dc_chatlist  dc_chatlist_t;
dc_chatlist_t* dc_get_chatlist(dc_context_t* context, int flags, const char* query_str, uint32_t query_id);
size_t    dc_chatlist_get_cnt(const dc_chatlist_t* chatlist);
uint32_t  dc_chatlist_get_chat_id(const dc_chatlist_t* chatlist, size_t index);
uint32_t  dc_chatlist_get_msg_id(const dc_chatlist_t* chatlist, size_t index);
dc_lot_t* dc_chatlist_get_summary(const dc_chatlist_t* chatlist, size_t index, dc_chat_t* chat);
void      dc_chatlist_unref(dc_chatlist_t* chatlist);
```

The majority of methods provide a pointer to a rust structure, which can be used to access its properties through specialized methods.
After using it you need to free it using the `_unref` methods (like `dc_chatlist_unref`), otherwise you will create memory leaks.

> ## <a id="why-implement-a-new-way"></a> Why implement a new way to talk to core? 

Reviewer 2: Maybe remove the `<a id ..>`? The TOC in the beginning is not rendering properly:

![image](https://github.com/user-attachments/assets/3e7f8f26-25a5-4228-b45a-bdd671811ecd)

> ## <a id="why-implement-a-new-way"></a> Why implement a new way to talk to core? 

Reviewer 2: ```suggestion
## Why implement a new way to talk to core? 
```

While using the cffi in android and iOS was working fine, in the desktop version which is based on electron it had some problems.

> The main problem was that Electron is a full browser which uses multiple processes, and you can't easily pass pointers to C-structs over process boundaries, ignoring that it is a bad idea. On Android and iOS you can just call Delta Chat core from the UI thread. So we ended up writing a JSON API on top of the Node.js NAPI bindings on top of the C bindings, more about that below in the comparison.

Reviewer 2: ```suggestion
The main problem was that Electron is a full browser which uses multiple processes, and you can't easily pass pointers to C-structs over process boundaries, ignoring that is a bad idea. On Android and iOS you can just call Delta Chat core from the UI thread. So we ended up writing a JSON API on top of the Node.js NAPI bindings on top of the C bindings, more about that below in the comparison.
```

---
title: Why we have 2 different ways to talk to core
author: simon
image: ../assets/blog/2024-11-10-why-jsonrpc-bindings-exist/jsonrpc-vs-cffi-thumbnail.png
# com_id:
render_toc: true
excerpt: This is a technical post about why we created the jsonrpc bindings additionally to our cffi bindings. Learn something about the internals of how Delta Chat is implemented.
---

> Foremost this is a quite technical post, read our other blog posts if you want to read something more targeted at end users.

If you have not yet looked at delta the delta chat source code, you might not know yet that we have a core library that is used by all UI implementation.
This means the core features like encryption, email protocols, chat and message management are using the same code on in all our apps (desktop, android, iOS).
This has the following benefits:

> - Because our Core library does all the heavy work and exposes easy methods such as `getAccounts`, `getChatlist`, `getChatContacts` and so on, it's much less work maintaining our apps on all platforms, because they basically only need to implement the UI.

Reviewer 3: ```suggestion
- Because the core library does all the heavy work and exposes easy methods such as `getAccounts`, `getChatlist`, `getChatContacts`, and so on, it's much less work maintaining the platform-specific apps, because they only need to implement the UI.
```

> - We have many tests for this core library, internal tests in rust and many integration tests in python.

Reviewer 3: ```suggestion
- The core library is thoroughly tested with rust unit tests and integration tests in Python that actually transfer e-mails.```

> - The core library can easily be used to write bots and new Apps/Clients (like deltatouch, a client for Ubuntu touch made by a community member in about a year, read the [blog post](./2023-07-02-deltatouch)).

Reviewer 3: ```suggestion
- The core library can easily be used to write bots and new apps/clients (like deltatouch, a client for Ubuntu touch made by a community member in about a year, read the [blog post](./2023-07-02-deltatouch)).
```

## The C Foreign Function Interface

The C Foreign Function Interface short CFFI was the first way to link to core.
It was introduced back when Björn started the Delta Chat project.
He wrote the [core in C](https://github.com/deltachat/deltachat-core) and forked the Signal android app for the UI, which is written in Java, so the core is connected via CFFI^[the header file [deltachat.h](https://github.com/deltachat/deltachat-core-rust/blob/main/deltachat-ffi/deltachat.h) is an easy way to get an idea of the API] and [JNI](https://github.com/deltachat/deltachat-android/blob/main/jni/dc_wrapper.c) (java native interface).

> Later [when we moved the core to rust](https://delta.chat/en/2019-05-08-xyiv#the-coming-delta-chat-rustocalypse) the CFFI stayed, and that the API stayed the same is also one of the reason why the migration from c to rust went so well.

Author: we could also call it ABI, that would be even more accurate

> Later [when we moved the core to rust](https://delta.chat/en/2019-05-08-xyiv#the-coming-delta-chat-rustocalypse) the CFFI stayed, and that the API stayed the same is also one of the reason why the migration from c to rust went so well.

Reviewer 2: `one of the reasons` in plural

The advantage of a CFFI is that most programming languages have some way to bind to it already built-in.

A peek into how methods in [`deltachat.h`](https://github.com/deltachat/deltachat-core-rust/blob/main/deltachat-ffi/deltachat.h) look like:

```c
char* dc_get_info(const dc_context_t* context);

#define DC_CONNECTIVITY_NOT_CONNECTED        1000
#define DC_CONNECTIVITY_CONNECTING           2000
#define DC_CONNECTIVITY_WORKING              3000
#define DC_CONNECTIVITY_CONNECTED            4000
int dc_get_connectivity(dc_context_t* context);

typedef struct _dc_chatlist  dc_chatlist_t;
dc_chatlist_t* dc_get_chatlist(dc_context_t* context, int flags, const char* query_str, uint32_t query_id);
size_t    dc_chatlist_get_cnt(const dc_chatlist_t* chatlist);
uint32_t  dc_chatlist_get_chat_id(const dc_chatlist_t* chatlist, size_t index);
uint32_t  dc_chatlist_get_msg_id(const dc_chatlist_t* chatlist, size_t index);
dc_lot_t* dc_chatlist_get_summary(const dc_chatlist_t* chatlist, size_t index, dc_chat_t* chat);

> void      dc_chatlist_unref(dc_chatlist_t* chatlist);

Reviewer 3: Yeah, this really is a bit complex.

```

The majority of methods provide a pointer to a rust structure, which can be used to access its properties through specialized methods.
After using it you need to free it using the `_unref` methods (like `dc_chatlist_unref`), otherwise you will create memory leaks.

## <a id="why-implement-a-new-way"></a> Why implement a new way to talk to core? 

While using the cffi in android and iOS was working fine, in the desktop version which is based on electron it had some problems.

> The main problem was that electron is a full browser which uses multiple processes, and you can't easily keep pass pointers to c-structs over process boundaries, ignoring that it is a bad idea. On android and iOS you can just call Delta Chat core from the UI thread. So we ended up writing a JSON API on top of the Node.js NAPI bindings on top of the c bindings, more about that below in the comparison.

Author: ignoring that passing pointers over process boundaries is not a good idea because it becomes even harder to keep track of and free them afterwards or you get massive memory leaks. Also if you pass pointers you might do it in a way where the pointers can get out of bounds - like client specifies another memory address it should not be able to access - if you pass the raw pointers as numbers. maybe this should be a footnote?

> The other problem in desktop that it is basically single threaded and while Delta Chat core uses async rust, the CFFI blocks on all calls (note the `block_on`):

Author: deltachats implementation of the cffi mainly: C does not have a concept of async coding, like there are no futures/promises built into the language. that and ease of user is why all calls that are asynchronous in the core are blocking in the cffi.

> The other problem in desktop that it is basically single threaded and while Delta Chat core uses async rust, the CFFI blocks on all calls (note the `block_on`):

Author: though all calls is not entirely true, there are some calls like `dc_configure` who are not blocking in the cffi because they spawn a thread, return immediately and you need to listen for events to find out when they are done and whether they were successful

```rust
pub unsafe extern "C" fn dc_stop_ongoing_process(context: *mut dc_context_t) {
  if context.is_null() {
    eprintln!("ignoring careless call to dc_stop_ongoing_process()");
    return;
  }
  let ctx = &*context;
  block_on(ctx.stop_ongoing());
}

> ```

Author: true there might be much better examples

> ```

Author: true there might be much better examples

In android and iOS you can easily start threads or defer blocking tasks to other threads, so there it is not a problem, but on desktop each call blocked the main process lead to an unresponsive experience.
Even though the communication between our main and UI process was already using async electron IPC, electron froze the UI process every time the main process was blocked.

Out of these and other frustrations the idea for a new way to talk to core was born. First there was only talk, like were discussions of what wire format to use: CBOR, message-pack or just plain JSON. A while nothing happened.

## The history of our jsonrpc interface

Then treefit started writing a deltachat-command-API project which was passing requests and answers over json.
There were two goals: make desktop development easier and to make the experiment of a KaiOS client possible [^kaios].

> [^kaios]: KaiOS has a similar problem: Only webapps are allowed, so there also a process boundary - KaiOS is small feature phones with T9 keyboard. BTW: treefit still plans to make that experimental client for KaiOS.

Reviewer 4: ```suggestion
[^kaios]: KaiOS has a similar problem: Only webapps are allowed, so there also is a process boundary - KaiOS is an OS for small feature phones with T9 keyboard. BTW: treefit still plans to make that experimental client for KaiOS.
```

After he got the [first prototype](https://github.com/Simon-Laux/delta-command-api) working, Frando cleaned up and rewrote the code to make it more [idiomatic and professional](https://github.com/deltachat/deltachat-jsonrpc/pull/14).
He also factored out the jsonrpc library and procedural macro into a dedicated rust crate, so that it can also be used by other projects: [yerpc](https://github.com/deltachat/yerpc)

Then we merged the repo into the core repo and moved desktop over to use the new jsonrpc API, which was easy thanks to the generated typescript bindings that gave good auto-completion.

> Though it still used the cffi and node bindings as transport, until May 17, 2024, when treefit migrated it to use the deltachat-rpc-server binary that uses stdio as a transport[^jsonrpc-pr].

Reviewer 4: Not super clear as transport between what in what. Maybe add something like "(see picture below)" if this is enough of an explanation?

[^jsonrpc-pr]: The pr: [use stdio binary instead of dc node & update electron to 30 & update min node to 20 #3567](https://github.com/deltachat/deltachat-desktop/pull/3567).

Desktop architecture versions:

<img src="../assets/blog/2024-11-10-why-jsonrpc-bindings-exist/mermaid-diagram-light.svg" data-dark-src="../assets/blog/2024-11-10-why-jsonrpc-bindings-exist/mermaid-diagram-dark.svg" />

When adding a new method we need to touch the code for the components shown in red, grey means that it is not touched, and green means the code is generated.

## The benefits of jsonrpc over CFFI

### Less Code to write - Simplify development

<img src="../assets/blog/2024-11-10-why-jsonrpc-bindings-exist/new-method-addtion-comparison.png" />

> 

Author: yes just to show difference

As you can see in this "meme", adding a method with jsonrpc is much simpler than adding a method to the CFFI.
This is thanks to 2 factors:

- A new method does not need to be specified at every stage. (see also the "Desktop architecture versions" diagram above)
- Code generation: typescript client wrapper code and [OpenRPC](https://open-rpc.org/) definition are auto generated from the rust code. (including documentation comments)

### Error handling

Error handling in c requires much discipline:
a common pattern in C is to return a status code and write results to a pointer that was provided as an argument to the function.

> Take this example from [libimobiledevice](https://libimobiledevice.org/):
> 
> > ```c
> > enum lockdownd_error_t {
> >   LOCKDOWN_E_SUCCESS = 0,
> >   LOCKDOWN_E_INVALID_ARG = -1,
> >   LOCKDOWN_E_INVALID_CONF = -2,
> >   LOCKDOWN_E_PLIST_ERROR = -3,
> >   LOCKDOWN_E_PAIRING_FAILED = -4,
> >   // ...
> > }
> > lockdownd_error_t lockdownd_client_new (idevice_t device, lockdownd_client_t *client, const char *label)
> > ```
> >
> > from <https://docs.libimobiledevice.org/libimobiledevice/latest/lockdown_8h.html>

Reviewer 4: ````suggestion
Take this example from [libimobiledevice](https://libimobiledevice.org/)'s [`lockdown.h`](https://docs.libimobiledevice.org/libimobiledevice/latest/lockdown_8h.html):

```c
enum lockdownd_error_t {
  LOCKDOWN_E_SUCCESS = 0,
  LOCKDOWN_E_INVALID_ARG = -1,
  LOCKDOWN_E_INVALID_CONF = -2,
  LOCKDOWN_E_PLIST_ERROR = -3,
  LOCKDOWN_E_PAIRING_FAILED = -4,
  // ...
}
lockdownd_error_t lockdownd_client_new (idevice_t device, lockdownd_client_t *client, const char *label)
```
````

Otherwise it's a scrollable region inside a scrollable region.

 *
 * @param data1 0
 * @param data2 (char*) Info string in English language.
 */
#define DC_EVENT_SMTP_CONNECTED           101
```

With the typescript bindings on the other hand you get named and typed properties.

Examples of events in jsonrpc:

```json
[
  {
    "kind": "SmtpMessageSent",
    "msg": "Message len=2402 was SMTP-sent to example@nine.testrun.org"
  },
  {
    "kind": "MsgDelivered",
    "chatId": 34,
    "msgId": 1342
  }
]
```

Usage in typescript:

```ts
dc.on("SmtpMessageSent", ({ msg }) => {
  console.log(msg);
});
dc.on("MsgDelivered", ({ chatId, msgId }) => {
  console.log(chatId, msgId);
});
```

This works very well with IDE auto-completion and "IntelliSense" (hover to get documentation).

### Async usage in Delta Chat desktop

As described in the [beginning](#why-implement-a-new-way), before jsonrpc we had issues with desktop getting unresponsive when there was a bit more api traffic, for example on fetching a lot of messages.
While the cffi is blocking, you never block on a single method call with jsonrpc since you just have one bidirectional stream of jsonrpc messages.
This really speed up desktop and made it more responsive.

> ### Easier to use than cffi over process boundaries

Author: is it better now?

> ⚠️ If you use websocket, be sure to use the encrypted `wss://` variant, because plain websockets `ws://` are as unencrypted as `http`

## Current Usage

Obviously we use this API in desktop, but we already use it to add new APIs also on iOS, android and for writing bots.
We have no autogenerated wrapper/client for any language besides typescript yet, so it is not as cool as it could be.
Though it is still easier to add new methods to jsonrpc than to the C-FFI, especially where core returns complex struct with many fields: In C-FFI you would need to create a new opaque struct, accessor functions for the properties and an unref function and wrappers for those functions; in jsonrpc you don't need all that, just write the rust struct & method and process the returned JSON in the client.

For this purpose, the C-FFI API actually contains the jsonrpc API and provides both synchronous and asynchronous access to it[^jsonrpc_in_cffi]. (These APIs were also used in desktop before we switched to the stdio-rpc-server binary.)

[^jsonrpc_in_cffi]: JSON-RPC API in the C-FFI: [`dc_jsonrpc_instance_t`](https://c.delta.chat/classdc__jsonrpc__instance__t.html)

> ## What's Next?

Reviewer 4: Maybe a "conclusion" section would be nice? Just to summarize the article, and let the reader make sure they haven't missed the point(s) and the author's vision. And just as a TL;DR.
Either above or below "What's next".

> Currently we only have autogeneration for the typescript bindings,

Reviewer 4: Auto-generation has been mentioned in the previous section.

Author: could you elaborate?

Reviewer 4: > We have no autogenerated wrapper/client for any language besides typescript yet

This is in the previous section. So repeating it here felt redundant, but IDK.

it would be great to also have autogenerated bindings for atleast Swift, Java and Python.
Also the documentation should be improved.
The code documentation part can easily be done by copying the great documentation from the cffi and adapting it to fit to the jsonrpc api
(this is a good-first-issue, if you, dear reader, want to help; improving documentation is easy to get into and is always welcome).

## Further reading

**Documentation:**

- <https://jsonrpc.delta.chat/>

> - list of all functions: <https://js.jsonrpc.delta.chat/classes/RawClient.html>

Reviewer 4: FYI This link refuses to be line-broken on narrow screens, which results in the page being horizontally scrollable.
